src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -503,6 +508,20 @@ private bool CompleteHandshake(ref ProtocolToken? alertToken, out SslPolicyError
                 return true;
             }
 
+#if TARGET_ANDROID
+            // If the remote certificate verification has already been invoked from Java TrustManager's callback
+            // we shouldn't run it repeatedly and honor the existing validation result.
+            // It's possible that the validation result isn't set (for example the client didn't send a certificate)
+            // in which case we still need to run the verification.
+            if (_securityContext?.ValidationResult is JavaProxy.RemoteCertificateValidationResult result)
```

Comment:
I'm not sure I fully understand the reasoning here. It may be worth to generalize it and wrap it into some PAL function. In general, if validation throws, the `AuthenticateAs*` would throw and it would normally cut the IO loop for handshake. I'm wondering how the Java is different.